src/Npgsql.OpenTelemetry/TracerProviderBuilderExtensions.cs

```diff
@@ -14,6 +15,12 @@ public static class TracerProviderBuilderExtensions
     /// </summary>
     public static TracerProviderBuilder AddNpgsql(
         this TracerProviderBuilder builder,
-        Action<NpgsqlTracingOptions>? options = null)
-        => builder.AddSource("Npgsql");
+        Action<NpgsqlTracingOptions>? configure = null)
+    {
+        var options = new NpgsqlTracingOptions();
+        configure?.Invoke(options);
+        return builder
+            .AddSource("Npgsql")
+            .AddInstrumentation(() => new NpgsqlTracingInstrumentation(options));
```

Comment:
I don't know the OpenTelemetry very well - what's the advantage of AddInstrumentation with NpgsqlTracingInstrumentation here over simply injecting NpgsqlActivitySource.Options directly in this method?

Reply:
I think we only want to set `NpgsqlActivitySource.Options` if and when `Build()` is called on the `TraceProviderBuilder`. I also think technically we want to remove those options once the built `TracerProvider` has been disposed. Using an instrumentation factory that returns an `IDisposable` gives us that behaviour.
(See 3.i [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#writing-own-instrumentation-library))

Reply:
Isn't that documentation section only relevant when the instrumentation library is separate/external, because the main library isn't instrumented? In other words, I think what we're doing qualifies as part of "make every library observable out of the box by having them call OpenTelemetry API directly" (the first sentence).

On a related note, I don't think this config stuff qualifies as "state management", which is what I think the AddInstrumentation API is about.

But I'm far from an expert on all this - let me know how you see things.

Reply:
I think regardless of if the documentation section is intended to be for just separate instrumentations, the behaviour it gives us is desirable here: only set `NpgsqlActivitySource.Options` if and when `Build()` is called on the `TraceProviderBuilder`, and unset it when the `TraceProvider` is disposed.